planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,339 @@
+import org.locationtech.jts.geom.Coordinate;
+import org.locationtech.jts.geom.CoordinateXY;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.GeometryComponentFilter;
+import org.locationtech.jts.geom.GeometryFactory;
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.PrecisionModel;
+
+public class LoopLineMerger {
+  private final List<LineString> input = new ArrayList<>();
+  private final List<Node> output = new ArrayList<>();
+  int ids = 0;
+  private PrecisionModel precisionModel = new PrecisionModel(16);
+  private GeometryFactory factory = new GeometryFactory(precisionModel);
+  private double minLength = 0.0;
+  private double loopMinLength = 0.0;
+
+  private static double getLength(List<Edge> lines) {
+    double result = 0.0;
+    for (var line : lines) {
+      result += line.length;
+    }
+    return result;
+  }
+
+  static boolean hasPointAppearingMoreThanTwice(List<Edge> edges) {
+    HashMap<Node, Integer> pointCountMap = new HashMap<>();
+    for (Edge edge : edges) {
+      if (pointCountMap.merge(edge.from, 1, Integer::sum) > 2) {
+        return true;
+      }
+      if (pointCountMap.merge(edge.to, 1, Integer::sum) > 2) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public LoopLineMerger setPrecisionModel(PrecisionModel precisionModel) {
+    this.precisionModel = precisionModel;
+    factory = new GeometryFactory(precisionModel);
+    return this;
+  }
+
+  public LoopLineMerger setMinLength(double minLength) {
+    this.minLength = minLength;
+    return this;
+  }
+
+  public LoopLineMerger setLoopMinLength(double loopMinLength) {
+    this.loopMinLength = loopMinLength;
+    return this;
+  }
+
+  public void add(Geometry geometry) {
+    geometry.apply((GeometryComponentFilter) component -> {
+      if (component instanceof LineString lineString) {
+        input.add(lineString);
+      }
+    });
+  }
+
+  private void merge() {
+    for (var node : output) {
+      if (node.getEdges().size() == 2) {
+        Edge a = node.getEdges().getFirst();
+        Edge b = node.getEdges().get(1);
+        node.getEdges().clear();
+        List<Coordinate> coordinates = new ArrayList<>();
+        coordinates.addAll(a.coordinates.reversed());
+        coordinates.addAll(b.coordinates.subList(1, b.coordinates.size()));
+        Edge c = new Edge(a.to, b.to, coordinates, a.length + b.length);
+        a.to.removeEdge(a.reversed);
+        b.to.removeEdge(b.reversed);
+        a.to.addEdge(c);
+        b.to.addEdge(c.reversed);
+      }
+    }
+  }
+
+  private void removeLoops() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        var allPaths = findAllPaths(edge.from, edge.to, loopMinLength);
+        if (allPaths.size() > 1) {
+          for (var path : allPaths.subList(1, allPaths.size())) {
+            for (var toRemove : path) {
+              toRemove.remove();
+            }
+          }
+        }
```

Comment:
I think this is wrong. For example consider the following graph:

![image](https://github.com/user-attachments/assets/290ecf05-39ea-4c1a-869b-4a742cb9d834)

When visiting the node B and the edge BD, then it would find the paths BD, BED, BCD and then it would remove BC, CD, BE, ED from the graph, which disconnects H and G from A and E. 

Maybe I can add a test for this.

Reply:
I fixed this now such that it corresponds to the old version which only removes the first edge of the non-shortest paths. I think there should be less line dropping now in openmaptiles because fewer islands are created.